src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -131,8 +128,38 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
             //       buffer on every render.
             var batchBytes = MessagePackSerializer.Serialize(batch, RenderBatchFormatterResolver.Instance);
 
+            if (_client.Connected)
+            {
+                // At this point, the server thinks that the client is connected and can be safely written to.
+                // If the client disconnects, writes using SignalR will fail silently so we don't need to do anything interesting here.
+                return WriteBatchBytes(batchBytes);
+            }
+            else
+            {
+                Log.SkippingRenderDisconnectedClient(_logger, CircuitId);
+
+                _disconnectedBatchBytes = batchBytes;
```

Comment:
I don't totally follow this. Why would there be exactly one render that occurs while disconnected? Is this array only used to capture the *first* render that is attempted following disconnection, with some different handling used for subsequent renders while disconnected?

There could be any number of renders attempted while disconnected (e.g., components updating on a timer). I don't yet know exactly what the preferred system for managing this would be. The two main approaches that come to mind are:

 * We could capture all of the `UpdateDisplay` payloads and attempt to dispatch them all (in order) on reconnection
 * Or, we could put the renderer into a mode where it doesn't actually perform renders while disconnected, but rather just tracks the list of components that will be rendered on the next reconnect. Then on reconnect, we do a single render batch that includes all those components.

My sense is that the second option will be better for various reasons.

It looks to me like neither of these (or any alternative) is implemented in this PR, but please let me know if I'm failing to spot how this is being handled!

Reply:
I agree with steve here. To shed some light on the scenario that steve just glanced at, imagine a component like the following

```
  public int Count { get; set; } = 0;
  
  OnInit()
  {
    StartTimer();
  }
  <p>Count: @Count</p>
  <ul>
  for(var i = 0; i < Count; i++){
    <li>i</li>
  }
  </ul>
```
With StartTimer being something that spins up a thread and calls `Component.Dispatch(() => Count++)` every second.

When I prerrender from lets say a Razor page, I want to render Count = 0 and no elements.
Then as the server is prerendering the server wants to save the initial prerrendered batch so that when the browser connects back from the client ("reconnect"), it can replace the content transparently, attach events, etc and initiate ("resume") interactivity. At that point, let say 10 seconds have passed. I expect to see Count = 10; And a list with 10 elements.

The way this is done right now, has a couple of caveheats. It won't work with prerrendered contents as the dom elements are not initialized with the correct state. That's why it's important to save and resend the first batch (because we essentially replace the contents of the initial DOM selector with the contents from the initial batch. (As an aside, I'm wondering if an optimization can be done here where we don't change the DOM and we just attach event handlers)

The important thing is that a render batch is the diff between the two renderbuildertree instances, not a complete snapshot. So if you loose a render batch you get into a corrupted state. In the prerendering case, we just simply don't get to even send the first one.

So as Steve above points out, you can do two things. 
* One is to save each serialized batch update and then resend them. I like this approach as it's dirt cheap and isolated to the remote renderer.
* Put the render into a special mode, similar to the one we do when we do DispatchEventAsync. This requires us to add things to the Renderer api, something like a Pause and Resume method, which I'm fine with, but it still requires you to capture that last render batch as one might be in progress when you decide to call Pause and that needs to finish.

I think I agree with steve in that the second approach feels better even though is a bit more complicated.

When it comes to prerrendering the way I see this working is as follows.
* Mvc starts prerrendering the component (we will add some timeout here so that we don't wait indefinitely)
  * After the time out happens or we produced a batch, the RemoteRenderer pauses the rendering and produces the initial batch (if we are not done) and the prerendered HTML conent.
* When components.server.js loads we scan the dom for prerrendered content markers:
  * For example <prerendered data-circuit-id="circuitidblob">
* components.server.js grabs the circuit ids on the page and reconnects to each of the circuits.
  * The initial (stored) batch is sent to the client and `<prerendered ...` and it's contents are replaced by the initial contents.
  * The circuit resumes sending updates and the client resumes sending events.

Reply:
I think we'll need to figure this out a bit better and I might need to dig in to how the rendering works. I changed this up to buffer content when we know the client is disconnected and resend them when the connection is back up. However it's possible that we may think we have sent several renders down to the client but they actual end up being sent to dev/null. e.g.

```
Connection
    | 1
    | 2
    | 3
Disconnected (client network error)
    | 4
    | 5
SignalR OnDisconnected is executed
   | 6 (buffered)
Client reconnected
```

We would buffer 6, but there could be several render diffs that were lost. Do we need a mechanism akin to `RenderRootComponentsAsync` that allows us to do this?

Reply:
@rynowak do you have any thoughts or suggestions here?

Reply:
For something like this to work, we either need to send a *full snapshot* when the client connects, or we need have an *ack* for render. I am not the best person to judge whether we currently have these capabilities.